File: src-tauri/src/coords.rs
//! 坐标转换工具
//! 支持 GCJ02 (高德) 和 BD09 (百度) 转 WGS84

use std::f64::consts::PI;

const X_PI: f64 = PI * 3000.0 / 180.0;
const A: f64 = 6378245.0;
const EE: f64 = 0.006_693_421_622_965_943;

/// BD09 坐标转 GCJ02
pub fn bd09_to_gcj02(bd_lon: f64, bd_lat: f64) -> (f64, f64) {
    let x = bd_lon - 0.0065;
    let y = bd_lat - 0.006;
    let z = (x * x + y * y).sqrt() - 0.00002 * (y * X_PI).sin();
    let theta = y.atan2(x) - 0.000003 * (x * X_PI).cos();
    let gcj_lon = z * theta.cos();
    let gcj_lat = z * theta.sin();
    (gcj_lon, gcj_lat)
}

/// GCJ02 坐标转 WGS84
pub fn gcj02_to_wgs84(gcj_lon: f64, gcj_lat: f64) -> (f64, f64) {
    if out_of_china(gcj_lon, gcj_lat) {
        return (gcj_lon, gcj_lat);
    }

    let dlat = transform_lat(gcj_lon - 105.0, gcj_lat - 35.0);
    let dlon = transform_lon(gcj_lon - 105.0, gcj_lat - 35.0);
    let radlat = gcj_lat / 180.0 * PI;
    let magic = radlat.sin();
    let magic = 1.0 - EE * magic * magic;
    let sqrtmagic = magic.sqrt();
    let dlat = (dlat * 180.0) / ((A * (1.0 - EE)) / (magic * sqrtmagic) * PI);
    let dlon = (dlon * 180.0) / (A / sqrtmagic * radlat.cos() * PI);
    (gcj_lon - dlon, gcj_lat - dlat)
}

/// BD09 坐标转 WGS84
pub fn bd09_to_wgs84(bd_lon: f64, bd_lat: f64) -> (f64, f64) {
    let (gcj_lon, gcj_lat) = bd09_to_gcj02(bd_lon, bd_lat);
    gcj02_to_wgs84(gcj_lon, gcj_lat)
}

/// 高德 GCJ02 坐标转 WGS84 (与 gcj02_to_wgs84 相同)
pub fn amap_to_wgs84(gcj_lon: f64, gcj_lat: f64) -> (f64, f64) {
    gcj02_to_wgs84(gcj_lon, gcj_lat)
}

fn out_of_china(lon: f64, lat: f64) -> bool {
    !(72.004..=137.8347).contains(&lon) || !(0.8293..=55.8271).contains(&lat)
}

fn transform_lat(x: f64, y: f64) -> f64 {
    let mut ret = -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * x.abs().sqrt();
    ret += (20.0 * (6.0 * x * PI).sin() + 20.0 * (2.0 * x * PI).sin()) * 2.0 / 3.0;
    ret += (20.0 * (y * PI).sin() + 40.0 * (y / 3.0 * PI).sin()) * 2.0 / 3.0;
    ret += (160.0 * (y / 12.0 * PI).sin() + 320.0 * (y * PI / 30.0).sin()) * 2.0 / 3.0;
    ret
}

fn transform_lon(x: f64, y: f64) -> f64 {
    let mut ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * x.abs().sqrt();
    ret += (20.0 * (6.0 * x * PI).sin() + 20.0 * (2.0 * x * PI).sin()) * 2.0 / 3.0;
    ret += (20.0 * (x * PI).sin() + 40.0 * (x / 3.0 * PI).sin()) * 2.0 / 3.0;
    ret += (150.0 * (x / 12.0 * PI).sin() + 300.0 * (x / 30.0 * PI).sin()) * 2.0 / 3.0;
    ret
}

File: src/lib/ais/coords.ts
// 中国坐标系转换（移植自 src-tauri/src/coords.rs，并补正向 WGS-84 → GCJ-02）。
// AIS 与水域面默认是 WGS-84；叠加到 GCJ-02 底图（高德/天地图）时需正向纠偏。

import type { BaseCrs, SourceCrs } from './types'

const PI = Math.PI
const X_PI = (PI * 3000) / 180
const A = 6378245.0
const EE = 0.006693421622965943

function outOfChina(lon: number, lat: number): boolean {
    return lon < 72.004 || lon > 137.8347 || lat < 0.8293 || lat > 55.8271
}

function transformLat(x: number, y: number): number {
    let ret = -100 + 2 * x + 3 * y + 0.2 * y * y + 0.1 * x * y + 0.2 * Math.sqrt(Math.abs(x))
    ret += ((20 * Math.sin(6 * x * PI) + 20 * Math.sin(2 * x * PI)) * 2) / 3
    ret += ((20 * Math.sin(y * PI) + 40 * Math.sin((y / 3) * PI)) * 2) / 3
    ret += ((160 * Math.sin((y / 12) * PI) + 320 * Math.sin((y * PI) / 30)) * 2) / 3
    return ret
}

function transformLon(x: number, y: number): number {
    let ret = 300 + x + 2 * y + 0.1 * x * x + 0.1 * x * y + 0.1 * Math.sqrt(Math.abs(x))
    ret += ((20 * Math.sin(6 * x * PI) + 20 * Math.sin(2 * x * PI)) * 2) / 3
    ret += ((20 * Math.sin(x * PI) + 40 * Math.sin((x / 3) * PI)) * 2) / 3
    ret += ((150 * Math.sin((x / 12) * PI) + 300 * Math.sin((x / 30) * PI)) * 2) / 3
    return ret
}

/** 计算 GCJ-02 相对 WGS-84 的偏移量（度） */
function gcjDelta(lon: number, lat: number): [number, number] {
    let dLat = transformLat(lon - 105, lat - 35)
    let dLon = transformLon(lon - 105, lat - 35)
    const radLat = (lat / 180) * PI
    let magic = Math.sin(radLat)
    magic = 1 - EE * magic * magic
    const sqrtMagic = Math.sqrt(magic)
    dLat = (dLat * 180) / (((A * (1 - EE)) / (magic * sqrtMagic)) * PI)
    dLon = (dLon * 180) / ((A / sqrtMagic) * Math.cos(radLat) * PI)
    return [dLon, dLat]
}

export function wgs84ToGcj02(lon: number, lat: number): [number, number] {
    if (outOfChina(lon, lat)) return [lon, lat]
    const [dLon, dLat] = gcjDelta(lon, lat)
    return [lon + dLon, lat + dLat]
}

export function gcj02ToWgs84(lon: number, lat: number): [number, number] {
    if (outOfChina(lon, lat)) return [lon, lat]
    const [dLon, dLat] = gcjDelta(lon, lat)
    return [lon - dLon, lat - dLat]
}

export function bd09ToGcj02(lon: number, lat: number): [number, number] {
    const x = lon - 0.0065
    const y = lat - 0.006
    const z = Math.sqrt(x * x + y * y) - 0.00002 * Math.sin(y * X_PI)
    const theta = Math.atan2(y, x) - 0.000003 * Math.cos(x * X_PI)
    return [z * Math.cos(theta), z * Math.sin(theta)]
}

export function bd09ToWgs84(lon: number, lat: number): [number, number] {
    const [gl, ga] = bd09ToGcj02(lon, lat)
    return gcj02ToWgs84(gl, ga)
}

/** 按源坐标系把入库点归一化到 WGS-84 */
export function normalizeToWgs84(crs: SourceCrs, lon: number, lat: number): [number, number] {
    if (crs === 'gcj02') return gcj02ToWgs84(lon, lat)
    if (crs === 'bd09') return bd09ToWgs84(lon, lat)
    return [lon, lat]
}

/** WGS-84 → 底图坐标系（用于渲染时对齐底图） */
export function toBase(base: BaseCrs, lon: number, lat: number): [number, number] {
    return base === 'gcj02' ? wgs84ToGcj02(lon, lat) : [lon, lat]
}
